### PR TITLE
DB-11930 Custom HBase Filter for Index Prefix Iteration

### PR DESCRIPTION
**Pull request by msirek**
## Short Description
Improve performance of Index Prefix Iteration with a large number of values in the first index column.

## Long Description
[SPLICE-2399](https://github.com/splicemachine/spliceengine/pull/5013) Added index key lookups with no predicates on the leading index column, improving query performance.  When the number of values in that column is high, the time taken to discover the values to probe is excessive.  For example, on a table with 130K values, it takes roughly 2 minutes to collect those 130K values.  The problem is that there is no way to dynamically alter a scan from the SpliceOperation side of execution (the SpliceOperation is converted to an HBase scan and fulfilled via a connection to a Region Server).  Once the first value is found, we can't modify the start key value (search key) of the scan, or modify a Filter attached to the scan, and have the modification take effect.  So, we have to close the scan and open a brand new scan with a different start key.  The creation and destruction of 130K scans is too costly, which accounts for the observed slowness.

The solution is to write a new custom HBase filter, named KeyPrefixProbingFilter, which knows how to find the length of the first column's constituent of the HBase rowkey (bytes in the HBase Cell) and probe within the first column values present.  Within a given first column value (in encoded format), a secondary MultiRowRangeFilter will be applied to probe the start keys and stop keys of index columns 2,3,4, etc.   Since this secondary filter needs to be reset and reapplied again and again, it needs special logic over the HBase MultiRowRangeFilter, which is single-pass.  A new SpliceMultiRowRangeFilter class which extends MultiRowRangeFilter is added for this purpose.  With these 2 new filters, the original costly value discovery logic is totally bypassed (except for the mem platform, which does not use HBase).  The query which previously took over 2 minutes now runs in under a second.

Also fixed in this pull request is a missing resetting of currentAccessPath.numUnusedLeadingIndexFields when costing the disableIndexPrefixIteration path, plus a reduction of optimizer costs for index prefix iteration plans when the favorIndexPrefixIteration session hint is used.

## How to test
connect 'jdbc:splice://localhost:1527/splicedb;user=splice;password=admin;useSpark=true';
CREATE TABLE FeatureStoreTest.FeatureStaging

(
    entity_key INTEGER,
    update_order INTEGER,
    ts TIMESTAMP,
    feature1 DOUBLE,
    feature2 DOUBLE,
    PRIMARY KEY (entity_key, update_order)
) ;

-- add entities at update_order = 1, with variable first timestamp
INSERT INTO FeatureStoreTest.FeatureStaging VALUES ( 1, 1, '2000-01-01 00:00:00', 0, 0 );

INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 1, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 2, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 4, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 8, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 16, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 32, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 64, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 128, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 256, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 512, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 1024, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 2048, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 4096, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 8192, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 16384, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 32768, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;
INSERT INTO FeatureStoreTest.FeatureStaging SELECT entity_key + 65536, update_order, timestampadd(SQL_TSI_DAY, CAST(ROUND(-44 + RANDOM()*88, 0) AS INTEGER), ts), feature1, feature2 FROM FeatureStoreTest.FeatureStaging;

INSERT INTO FeatureStoreTest.FeatureStaging
SELECT entity_key,
       update_order + 1,
       timestampadd(SQL_TSI_DAY, CAST( randValue AS INTEGER), ts),
       feature1 + randValue,
       feature2 + randValue
FROM FeatureStoreTest.FeatureStaging,
     (VALUES(CAST(ROUND(-4 + RANDOM()*8, 0) AS INTEGER)))oneRandom(randValue);

INSERT INTO FeatureStoreTest.FeatureStaging
SELECT entity_key,
       update_order + 2,
       timestampadd(SQL_TSI_DAY, CAST( randValue AS INTEGER), ts),
       feature1 + randValue,
       feature2 + randValue
FROM FeatureStoreTest.FeatureStaging,
     (VALUES(CAST(ROUND(-4 + RANDOM()*8, 0) AS INTEGER)))oneRandom(randValue);

INSERT INTO FeatureStoreTest.FeatureStaging
SELECT entity_key,
       update_order + 4,
       timestampadd(SQL_TSI_DAY, CAST( randValue AS INTEGER), ts),
       feature1 + randValue,
       feature2 + randValue
FROM FeatureStoreTest.FeatureStaging,
     (VALUES(CAST(ROUND(-4 + RANDOM()*8, 0) AS INTEGER)))oneRandom(randValue);
 
INSERT INTO FeatureStoreTest.FeatureStaging
SELECT entity_key,
       update_order + 8,
       timestampadd(SQL_TSI_DAY, CAST( randValue AS INTEGER), ts),
       feature1 + randValue,
       feature2 + randValue
FROM FeatureStoreTest.FeatureStaging,
     (VALUES(CAST(ROUND(-4 + RANDOM()*8, 0) AS INTEGER)))oneRandom(randValue);

analyze table FeatureStaging;
call syscs_util.syscs_flush_table('FEATURESTORETEST', 'FEATURESTAGING');

/* The following SQL should take about 1 second to run.  Before it took 2-3 minutes. */

> explain
> SELECT count(*)
> FROM FeatureStoreTest.FeatureStaging  
> WHERE update_order=1;
> Plan
> //----
> Cursor(n=5,rows=1,updateMode=READ_ONLY (1),engine=OLAP (cost))
>   ->  ScrollInsensitive(n=5,totalCost=4273.544,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3)
>     ->  ProjectRestrict(n=4,totalCost=244.25,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3)
>       ->  GroupBy(n=3,totalCost=244.25,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3)
>         ->  ProjectRestrict(n=2,totalCost=732.699,outputRows=131072,outputHeapSize=2.1 MB,partitions=1,parallelTasks=3)
>           ->  ProjectRestrict(n=1,totalCost=732.699,outputRows=131072,outputHeapSize=2.1 MB,partitions=1,parallelTasks=3)
>             ->  TableScan[FEATURESTAGING(1696),IndexPrefixIteratorMode(132480 values)](n=0,totalCost=732.699,scannedRows=131072,outputRows=131072,outputHeapSize=2.1 MB,partitions=1,parallelTasks=3,keys=[(UPDATE_ORDER[0:2] = 1)])
> 
> 
> SELECT count(*)
> FROM FeatureStoreTest.FeatureStaging 
> WHERE update_order=1;
> 
> splice> > > 1
> //-
> 131072
> 
> 1 row selected
> ELAPSED TIME = 822 milliseconds

### CODE REVIEW

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3006/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

**Comment by msirek** [comment]
jenkins please test branch @dbaas3.0

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

File: hbase_storage/pom.xml [hunk 1]
```diff
@@ -27,6 +27,11 @@
             <artifactId>pipeline_api</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.splicemachine</groupId>
+            <artifactId>splice_machine</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (dgomezferro) [inline hbase_storage/pom.xml:34]: Why is this needed? In principle I'd say `hbase_storage` shouldn't depend on `splice_machine`, since `splice_machine` should be a higher level module.

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static com.splicemachine.db.shared.common.reference.SQLState.LANG_INTERNAL_ERROR;
+
+@SuppressFBWarnings(value="HE_EQUALS_NO_HASHCODE", justification="Intentional")
+public class SpliceMultiRowRangeFilter extends MultiRowRangeFilter {
```
Comment (dgomezferro) [inline hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java:41]: Can you add some high-level documentation to the class explaining the difference vs MultiRowRangeFilter ?

File: hbase_storage/src/main/java/com/splicemachine/storage/HScan.java [hunk 3]
```diff
@@ -240,4 +249,11 @@ public Scan unwrapDelegate(){
     public void setSmall(boolean small) {
         scan.setSmall(small);
     }
+
+    @Override
+    public void attachKeyPrefixFilter(DataValueDescriptor firstColumn, String tableVersion) throws IOException {
+        Filter secondaryFilter = scan.getFilter();
```
Comment (hatyo) [inline hbase_storage/src/main/java/com/splicemachine/storage/HScan.java:255]: Is it possible that we get e.g. a `FilterList` here instead of `MultiRowRangeFilter`? Could you please add some checks for this?

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    public Cell getNextCellHint(Cell currentKV) {
+        // skip to the next range's start row
+        // #getComparisonData lets us avoid the `if (reversed)` branch
+        byte[] comparisonData = range.getComparisonData();
+        return PrivateCellUtil.createFirstOnRow(comparisonData, 0, (short) comparisonData.length);
+    }
+
+    @Override
+    public boolean filterRowKey(Cell firstRowCell) {
+        if (filterAllRemaining()) return true;
```
Comment (hatyo) [inline hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java:85]: this is superfluous since `filterAllRemaining` always returns false.

File: hbase_storage/pom.xml (hunk 1 above)
Comment (msirek) [reply hbase_storage/pom.xml:34]: It's needed for DataValueDescriptor in KeyPrefixProbingFilter.  I've refactored that class so that it's no longer needed.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/iapi/ScanInformation.java [hunk 5]
```diff
@@ -65,13 +65,15 @@
     DataScan getScan(TxnView txn, T startKeyHint, int[] keyDecodingMap, T stopKeyPrefix,
                      List<Pair<T, T>> keyRows, DataValueDescriptor scanKeyPrefix,
                      boolean sameStartStopScanKeyPrefix,
-                     List<ExecRow> firstIndexColumnKeys) throws StandardException;
+                     List<ExecRow> firstIndexColumnKeys,
+                     boolean SkipBuildOfFirstKeyColumn) throws StandardException;
```
[the remark this answers is not in the document]
Comment (msirek) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/iapi/ScanInformation.java:69]: Made the change

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java (hunk 2 above)
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java:41]: Added javadocs.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java [hunk 6]
```diff
@@ -5104,4 +5104,15 @@ public OptimizablePredicate getUsefulPredicateForUnionedIndexScan(FromBaseTable
         return null;
     }
 
+    public boolean hasLeadingIndexColumnStartOrStopKey() {
```
[the remark this answers is not in the document]
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java:5107]: Predicates are pushed to and pulled from PredicateLists all the time, so this is not a constant property which could be cached.  In any case, this method is not used in the latest implementation.

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import static com.splicemachine.encoding.MultiFieldDecoder.create;
+import static org.apache.hadoop.hbase.filter.Filter.ReturnCode.INCLUDE;
+import static org.apache.hadoop.hbase.filter.Filter.ReturnCode.SEEK_NEXT_USING_HINT;
+
+public class KeyPrefixProbingFilter extends FilterBase implements Writable{
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:46]: Added javadocs

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    protected Filter secondaryFilter;
+
+    public KeyPrefixProbingFilter(){
+        super();
+    }
+
+    public KeyPrefixProbingFilter(DataValueDescriptor firstKeyColumnDVD, String tableVersion, Filter secondaryFilter){
+        this();
+        if (firstKeyColumnDVD == null)
+            throw new RuntimeException();
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:72]: This parameter is removed entirely now after some refactoring.

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        if (tableVersion == null)
+            throw new RuntimeException();
+        if (!(secondaryFilter instanceof MultiRowRangeFilter))
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:75]: Sure.  Made the change.  If it supports more filters in the future we can create a new constructor.

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+    public byte [] getFirstKeyColumnDVDAsByteArray() {
+        return firstKeyColumnDVDAsByteArray;
+    }
+
+    public String getTableVersion() {
+        return tableVersion;
+    }
+
+    @SuppressFBWarnings(value = "URF_UNREAD_FIELD", justification = "intentional")
+    private void populateSerializers() {
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:109]: This method is removed after refactoring.

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+            secondaryFilter = SpliceMultiRowRangeFilter.parseFrom(filterAsBytes);
+        }
+        catch (DeserializationException de) {
+            throw new IOException(de);
+        }
+        populateSerializers();
+    }
+
+    @Override
+    public ReturnCode filterCell(final Cell ignored) throws IOException {
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:153]: The interface of org.apache.hadoop.hbase.filter.Filter requires a Cell to be passed to filterCell, but the determination of the ReturnCode to use was actually made in filterRowKey.  However, filterRowKey does not return a ReturnCode.  It returns a boolean, so this is like a deferred communication of the ReturnCode to HBase.  First filterRowKey determines the ReturnCode.  Then filterRowKey returns false, indicating filterCell should be called.  Then filterCell returns the stored ReturnCode (indicating if we should include the current Cell/rowkey, or if we should skip forward).

File: hbase_storage/src/main/java/com/splicemachine/storage/HScan.java (hunk 3 above)
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/storage/HScan.java:255]: This is not supposed to happen.  I've added the following here:

>         if (!(secondaryFilter instanceof MultiRowRangeFilter))
>             throw new IOException(StandardException.newException(LANG_INTERNAL_ERROR,
>                     "Expected KeyPrefixFilter's secondary filter to be a MultiRowRangeFilter."));

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java (hunk 4 above)
Comment (msirek) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/SpliceMultiRowRangeFilter.java:85]: True, but if the implementation is modified in the future, this would be required for peak performance.  It's good to leave this in so it's not something that has to be remembered or rediscovered.

File: hbase_storage/pom.xml (hunk 1 above)
Comment (msirek) [reply hbase_storage/pom.xml]: I've added a benchmark test.

**Comment by msirek** [comment]
jenkins please test branch @dbaas3.0

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3046/

**Comment by msirek** [comment]
jenkins please test branch @dbaas3.0

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3051/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0

**Review by dgomezferro — APPROVED** [review]

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

File: hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (hatyo) [reply hbase_storage/src/main/java/com/splicemachine/derby/hbase/KeyPrefixProbingFilter.java:131]: Thanks for the detailed explanation!

**Review by hatyo — APPROVED** [review]

**Review by carolp-503 — COMMENTED** [review]
+1
